[webkitbugspy] Add support for sourceChanges

JonWBedard opened the pull request:
#### 1517cb01bf1f338b0486f5c72ac9c3a02f784524
<pre>
[webkitbugspy] Add support for sourceChanges
<a href="https://bugs.webkit.org/show_bug.cgi?id=254023">https://bugs.webkit.org/show_bug.cgi?id=254023</a>
<a href="https://rdar.apple.com/106808712">rdar://106808712</a>

Reviewed by Dewei Zhu.

Add tooling to support the &quot;sourceChanges&quot; field in Radar, specify that field
as unsupported in GitHub and Bugzilla.

* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.populate): Populate source_changes with an empty list.
(Tracker.set): Reject any attempts to modify the unsupported source_changes field.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
(Tracker.populate): Populate source_changes with an empty list.
(Tracker.set): Reject any attempts to modify the unsupported source_changes field.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py:
(Issue.__init__): Add _source_changes field.
(Issue.source_changes): Return _source_changes.
(Issue.add_source_change): Add a source change to the source_changes field.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
(RadarModel.__init__): Add sourceChanges field, if specified.
(RadarModel.commit_changes): Commit sourceChanges, if available.
(RadarClient.radar_for_id): Pass additional fields to RadarModel constructor.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker.populate): Parse and populate source_changes.
(Tracker.set): Construct and update sourceChange field.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/github_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:

Canonical link: <a href="https://commits.webkit.org/277553@main">https://commits.webkit.org/277553@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e93c018d5bff19aa6974b1ea54b9d8459fb4afc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47181 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26360 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49825 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49864 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43230 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49488 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31689 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23819 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38432 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47762 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/23848 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40670 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19743 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/47045 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21257 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/41826 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5225 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/43561 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/42234 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51740 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/22208 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18584 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45726 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23483 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44733 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/24269 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6800 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23200 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

akath20 commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -350,6 +351,10 @@ def set(self, issue, assignee=None, opened=None, why=None, project=None, compone
 
+        if source_change:
+            sys.stderr.write('Bugzilla does not support source changes at this time\n')
+            return None
```
Does this make sense to `return None` in this case? Seems like it could act more as a warning instead of somewhat of a "fatal" operation to this command by not returning the issue back

JonWBedard replied:
My thinking here is that if we don't return `None`, a program wouldn't be able to tell that its request to `set` the `source_change` failed. It may be the case that program doesn't do anything with the success/failure state (and right now, no program is using this particular code path), but it seems like we should at least make it detectable? Especially in the case where the program is actually a service whose logs we rarely inspect.

akath20 commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
```diff
@@ -159,6 +159,9 @@ def __init__(self, client, issue):
         else:
             self.component = None
 
+        if 'sourceChanges' in (additional_fields or []):
```
Nit: although a test mock test case, would be nice to set `additional_fields = additional_field or []` right away in case it's used later on.

dewei-zhu reviewed (approved):
r=me

JonWBedard committed:
[webkitbugspy] Add support for sourceChanges
https://bugs.webkit.org/show_bug.cgi?id=254023
rdar://106808712

Reviewed by Dewei Zhu.

Add tooling to support the "sourceChanges" field in Radar, specify that field
as unsupported in GitHub and Bugzilla.

* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.populate): Populate source_changes with an empty list.
(Tracker.set): Reject any attempts to modify the unsupported source_changes field.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
(Tracker.populate): Populate source_changes with an empty list.
(Tracker.set): Reject any attempts to modify the unsupported source_changes field.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py:
(Issue.__init__): Add _source_changes field.
(Issue.source_changes): Return _source_changes.
(Issue.add_source_change): Add a source change to the source_changes field.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
(RadarModel.__init__): Add sourceChanges field, if specified.
(RadarModel.commit_changes): Commit sourceChanges, if available.
(RadarClient.radar_for_id): Pass additional fields to RadarModel constructor.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker.populate): Parse and populate source_changes.
(Tracker.set): Construct and update sourceChange field.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/github_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:

Canonical link: https://commits.webkit.org/277553@main
